### PR TITLE
Build ebpfcore as a DLL for testing

### PR DESCRIPTION
**Pull request by dthaler**
## Description

* Update to latest usersim
* Fix bug in libbpf_link.cpp found by latest usersim, where `ebpf_free()` was being used to free memory allocated by `strdup()` rather than an ebpf allocator
* Fix memory leaks in libbpf_program.cpp where `strdup()` was used to allocate memory that was never freed
* Fix bugs in test_helper.cpp found by latest usersim, where `ebpf_free()` was being used to free memory allocated by `new`.
* Fixed bug found by latest usersim, where ebpf_program.c used `ebpf_free()` to free memory allocated by ubpf_user.c using `calloc()`.  ubpf_kernel.c already redefined it to use ebpf_alloc but ubpf_user.c didn't match.
* Fix mismatched alloc/free in verifier_fuzzer found by usersim.  The verifier_fuzzer called an API that allocated memory using `ebpf_allocate()` but verifier_fuzzer then called `free()`
* Build ebpfcore_usersim.dll using the same files as used for ebpfcore.sys.
* Add a test case that exercises the DriverEntry/DriverUnload code paths.

## Testing

Tests will be expanded in the future once IOCTL support is added to usersim.

## Documentation

No impact.

## Installation

Installer files are updated in this PR.

### CODE REVIEW

**Comment by dthaler** [comment]
Blocked on https://github.com/microsoft/usersim/pull/80

**Comment by dthaler** [comment]
Blocked on https://github.com/iovisor/ubpf/pull/335

File: libs/api/libbpf_link.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@ bpf_link__pin(struct bpf_link* link, const char* path)
         return libbpf_err(-EBUSY);
     }
 
-    link->pin_path = strdup(path);
+    link->pin_path = ebpf_strdup(path);
```
Comment (saxena-anurag) [inline libs/api/libbpf_link.cpp:18]: `strdup()` internally calls `malloc()`. Before the usersim changes, the user mode implementation of `ebpf_free()` called `free()`, which was correct, as it matched with the corresponding `malloc()` in strdup(). 

Now it looks like there is only one implementation of `ebpf_free()`, both for user mode and kernel mode, and it calls `ExFreePool()`. So, for user mode code also, we are invoking user sim implementation of `ExFreePool()`.

Is this expected?

Comment (dthaler) [reply libs/api/libbpf_link.cpp:18]: Until #2677 is done, yes this is expected.

File: ebpfsvc/rpc_api.cpp [hunk 2]
```diff
@@ -36,8 +40,19 @@ ebpf_server_verify_and_load_program(
...
         reinterpret_cast<ebpf_inst*>(info->instructions),
-        const_cast<const char**>(logs),
-        logs_size);
+        &ebpf_logs,
+        &ebpf_logs_size);
+
+    if (ebpf_logs) {
+        // The ebpf_logs buffer was allocated by the ebpf allocator whereas we
+        // must return a string allocated by the MIDL allocator.
+        *logs = (char*)MIDL_user_allocate(ebpf_logs_size);
```
Comment (saxena-anurag) [inline ebpfsvc/rpc_api.cpp:49]: It looks like there are 2 versions of `MIDL_user_allocate()` currerntly in the repo. The client version is calling `ebpf_allocate()`, whereas the server version is calling `malloc()`.

If we change the implementation of the server version to also call `ebpf_allocate()`, then do we still need this change?

File: libs/api/libbpf_program.cpp [hunk 3]
```diff
@@ -390,7 +399,12 @@ bpf_object__pin_programs(struct bpf_object* obj, const char* path)
         char buf[PATH_MAX];
         int len;
 
-        len = snprintf(buf, PATH_MAX, "%s/%s", path, __bpf_program__pin_name(prog));
+        char* pin_name = __bpf_program__pin_name(prog);
```
Comment (saxena-anurag) [inline libs/api/libbpf_program.cpp:402]: unrelated to changes in this PR: we are allocating memory in failure path. Instead should we instead save the pin paths generated when pinning the programs, and use the same in the failure case?

File: ebpfsvc/rpc_api.cpp (hunk 2 above)
Comment (dthaler) [reply ebpfsvc/rpc_api.cpp]: Done

**Review by saxena-anurag — APPROVED** [review]

**Review by saxena-anurag — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**